### PR TITLE
Add support for audit logging inside user clusters

### PR DESCRIPTION
**Pull request by eqrx**
**What this PR does / why we need it**:
Requirement for [this issue](https://app.zenhub.com/workspaces/kubermatic-customer-5caccadc8a630d4a65e9590b/issues/kubermatic/customer-projects/158).

**Special notes for your reviewer**:
None

**Documentation**:
No changes yet.

**Does this PR introduce a user-facing change?**:
```release-note
Add support for kubernetes auditing
```

### CODE REVIEW

File: api/pkg/resources/apiserver/deployment.go [hunk 1]
```diff
@@ -37,6 +37,17 @@ const (
 	defaultNodePortRange = "30000-32767"
 )
 
+func AuditConfigMapCreator() reconciling.NamedConfigMapCreatorGetter {
+	return func() (string, reconciling.ConfigMapCreator) {
+		return resources.AuditConfigMapName, func(cm *corev1.ConfigMap) (*corev1.ConfigMap, error) {
+			if cm.Data == nil {
+				cm.Data = map[string]string{"audit.yaml": "apiVersion: audit.k8s.io/v1\nkind: Policy\nrules:\n- level: Metadata\n"}
```
Comment (mrIncompetent) [inline api/pkg/resources/apiserver/deployment.go]: Make this a multi line string to aid readability

File: api/pkg/resources/apiserver/deployment.go [hunk 2]
```diff
@@ -162,6 +173,23 @@ func DeploymentCreator(data *resources.TemplateData, enableDexCA bool) reconcili
 				},
 			}
 
+			dep.Spec.Template.Spec.Containers = append(dep.Spec.Template.Spec.Containers,
+				corev1.Container{
+					Name:    "audit-stdout",
```
Comment (mrIncompetent) [inline api/pkg/resources/apiserver/deployment.go]: `audit-logs` maybe?

File: api/pkg/resources/apiserver/deployment.go [hunk 3]
```diff
@@ -162,6 +173,23 @@ func DeploymentCreator(data *resources.TemplateData, enableDexCA bool) reconcili
...
+					Image:   "fluent/fluent-bit:1.0",
```
Comment (mrIncompetent) [inline api/pkg/resources/apiserver/deployment.go]: v1.2.2 is already available & is what we use for the FluentBit chart

File: api/pkg/resources/apiserver/deployment.go [hunk 4]
```diff
@@ -162,6 +173,23 @@ func DeploymentCreator(data *resources.TemplateData, enableDexCA bool) reconcili
...
+					Command: []string{"/fluent-bit/bin/", "fluent-bit"},
+					Args:    []string{"-i", "tail", "-p", "path=/var/log/kubernetes/audit/audit.log", "-p", "db=/var/log/kubernetes/audit/fluentbit.db", "-o", "stdout"},
+					VolumeMounts: []corev1.VolumeMount{
+						{
+							Name:      resources.AuditLogVolumeName,
+							MountPath: "/var/log/kubernetes/audit",
+							ReadOnly:  false,
+						},
+					},
+					Resources: *defaultResourceRequirements.DeepCopy(),
```
Comment (mrIncompetent) [inline api/pkg/resources/apiserver/deployment.go]: Specify custom resource requirements for this container - The default ones are for the apiserver container and are way too big

File: api/pkg/resources/apiserver/deployment.go [hunk 5]
```diff
@@ -255,6 +283,8 @@ func getApiserverFlags(data *resources.TemplateData, etcdEndpoints []string, ena
 		}
 	}
 
+	flags = append(flags, "--audit-policy-file", "/etc/kubernetes/audit/policy.yaml")
```
Comment (mrIncompetent) [inline api/pkg/resources/apiserver/deployment.go]: Make this part of the slice initialization & put it to the other audit flags

File: api/pkg/resources/apiserver/deployment.go [hunk 6]
```diff
@@ -332,11 +362,24 @@ func getVolumeMounts(enableDexCA bool) []corev1.VolumeMount {
 		})
 	}
 
+	volumesMounts = append(volumesMounts,
```
Comment (mrIncompetent) [inline api/pkg/resources/apiserver/deployment.go]: Put this to the slice intialization

File: api/pkg/resources/apiserver/deployment.go [hunk 7]
```diff
@@ -434,6 +477,28 @@ func getVolumes() []corev1.Volume {
 			},
 		},
 	}
+
+	volumes = append(volumes,
```
Comment (mrIncompetent) [inline api/pkg/resources/apiserver/deployment.go]: Put it to the slice initialization

File: api/pkg/resources/apiserver/deployment.go [hunk 8]
```diff
@@ -162,6 +173,23 @@ func DeploymentCreator(data *resources.TemplateData, enableDexCA bool) reconcili
 				},
 			}
 
+			dep.Spec.Template.Spec.Containers = append(dep.Spec.Template.Spec.Containers,
+				corev1.Container{
+					Name:    "audit-stdout",
+					Image:   "fluent/fluent-bit:1.0",
+					Command: []string{"/fluent-bit/bin/", "fluent-bit"},
```
Comment (mrIncompetent) [inline api/pkg/resources/apiserver/deployment.go]: Whats the advantage of specifying 2 instead of a single `/fluent-bit/bin/fluent-bit` ?

Comment (eqrx) [reply api/pkg/resources/apiserver/deployment.go]: Line 140 uses this as apiserver command. Wanted to be consistent, no idea why this is set that way

**Pull request merged**